sqlglot/optimizer/qualify_columns.py

```diff
@@ -528,61 +539,82 @@ def get_table(self, column_name: str) -> t.Optional[exp.Identifier]:
         return exp.to_identifier(table_name)
 
     @property
-    def all_columns(self):
+    def all_columns(self) -> t.Set[str]:
         """All available columns of all sources in this scope"""
         if self._all_columns is None:
             self._all_columns = {
                 column for columns in self._get_all_source_columns().values() for column in columns
             }
         return self._all_columns
 
-    def get_source_columns(self, name, only_visible=False):
-        """Resolve the source columns for a given source `name`"""
+    def get_source_columns(
+        self,
+        name: str,
+        source: t.Optional[exp.Expression | Scope] = None,
+        only_visible: bool = False,
+    ) -> t.List[str]:
+        """Resolve the source columns for a given source `name`."""
         if name not in self.scope.sources:
             raise OptimizeError(f"Unknown table: {name}")
 
-        source = self.scope.sources[name]
+        original_source = self.scope.sources[name]
 
-        # If referencing a table, return the columns from the schema
-        if isinstance(source, exp.Table):
-            return self.schema.column_names(source, only_visible)
+        if isinstance(original_source, exp.Table):
+            # If referencing a table, return the columns from the schema
+            columns = self.schema.column_names(original_source, only_visible)
+        elif isinstance(original_source, Scope) and isinstance(
+            original_source.expression, exp.Values
+        ):
+            columns = original_source.expression.alias_column_names
+        else:
+            # Otherwise, if referencing another scope, return that scope's named selects
+            columns = original_source.expression.named_selects
 
-        if isinstance(source, Scope) and isinstance(source.expression, exp.Values):
-            return source.expression.alias_column_names
+        if isinstance(source, Scope):
+            column_aliases = source.expression.alias_column_names
+        elif isinstance(source, exp.Expression):
+            column_aliases = source.alias_column_names
+        else:
+            column_aliases = []
 
-        # Otherwise, if referencing another scope, return that scope's named selects
-        return source.expression.named_selects
+        return [alias or name for (name, alias) in itertools.zip_longest(columns, column_aliases)]
 
-    def _get_all_source_columns(self):
+    def _get_all_source_columns(self) -> t.Dict[str, t.List[str]]:
         if self._source_columns is None:
             self._source_columns = {
-                k: self.get_source_columns(k)
-                for k in itertools.chain(self.scope.selected_sources, self.scope.lateral_sources)
+                name: self.get_source_columns(name, source=seq_get(ensure_list(source), 0))
```

Comment:
why are you doing ensure_list seq_get?

Reply:
`selected_sources` is a dict from `str` to a `tuple`, from which we want the first element. On the other hand, `lateral_sources` maps a `str` directly to a `Table | Scope` object.

The motivation is to always get the source code, regardless of whether we're examining a `selected_source` or a `lateral_source`.